### PR TITLE
fix(config): parse wrapped didChange settings

### LINKED ISSUE #595: workspace/didChangeConfiguration: section-wrapped payloads silently no-op (and typod keys are silently accepted) while logging "Configuration updated!"
## Found by config/CLI audit + codex cross-check

src/lsp/lsp_impl/workspace/did_change_configuration.rs:27-58 —

1. **Section-wrapped payloads no-op.** Clients commonly send `{"settings": {"kakehashi": {...}}}` (section-wrapped) rather than the flat shape. The handler deserializes only the flat shape; a section-wrapped payload deserializes to an effectively empty update, is merged without effect, and the handler still logs "Configuration updated!". Users on such clients cannot change settings mid-session and get positive feedback that they did.

2. **Unknown fields silently accepted.** `RawWorkspaceSettings` lacks `deny_unknown_fields` (or an unknown-key warning), so a misspelled key produces the same silent-success behavior.

### Fix direction
Try the section-wrapped shape when flat deserialization yields an empty object (or check for a `kakehashi` key first); log a warning listing ignored/unknown keys instead of unconditional success.

### PR DESCRIPTION
**Pull request by atusy**
## Summary
- parse `workspace/didChangeConfiguration` payloads wrapped as `settings.kakehashi` or `kakehashi`
- preserve deprecation scanning on the unwrapped kakehashi config value
- warn when client config contains ignored unknown top-level keys

Fixes #595

## Tests
- `cargo test lsp::lsp_impl::workspace::did_change_configuration::tests`
- `cargo fmt --check`
- `cargo test --test e2e_config_file did_change` (0 tests matched in this build)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of workspace configuration updates by normalizing nested or partially provided settings before applying them.
  * Added more reliable warnings for unknown configuration keys and better detection of “effectively empty” updates to avoid unintended changes.
  * Preserved deprecated `rootMarkers` settings so prior behavior remains consistent.

* **Tests**
  * Added unit tests covering top-level vs section-wrapped parsing, settings merging, unknown-key warnings, effectively empty updates (including the `languageServers` edge case), and deprecated-field preservation.

### CODE REVIEW

**Comment by atusy** [comment]
@revise

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 1]
```diff
@@ -68,3 +72,99 @@ impl Kakehashi {
...
+    let raw_value = unwrap_kakehashi_settings(value);
+    let warnings = ignored_key_warnings(&raw_value);
+    let settings = serde_json::from_value::<RawWorkspaceSettings>(raw_value.clone())?;
+
+    Ok(ClientConfigurationUpdate {
+        settings,
+        raw_value,
+        warnings,
+    })
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:112]: Fixed in 9e816b92 and 234ae891. The normalizer now treats wrapped settings from other servers as an empty update unless there is an actual kakehashi object or known kakehashi configuration key, so unrelated editor/server keys no longer produce unknown-key warnings. Added coverage for settings-root keys and unrelated gopls settings.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 2]
```diff
@@ -68,3 +72,99 @@ impl Kakehashi {
...
+
+fn unwrap_kakehashi_settings(value: serde_json::Value) -> serde_json::Value {
+    value
+        .get("settings")
+        .and_then(|settings| settings.get("kakehashi"))
+        .cloned()
+        .or_else(|| value.get("kakehashi").cloned())
+        .unwrap_or(value)
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 9e816b92. normalize_kakehashi_settings now only unwraps kakehashi values that are JSON objects; non-object kakehashi values are treated as empty updates instead of causing RawWorkspaceSettings deserialization errors. Added non_object_kakehashi_settings_are_empty coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 3]
```diff
@@ -68,3 +72,99 @@ impl Kakehashi {
...
+            "autoInstall": true
+        }))
+        .expect("settings with unknown keys should still parse");
+
+        assert_eq!(update.settings.auto_install, Some(true));
+        assert_eq!(
+            update.warnings,
+            vec!["Ignored unknown client configuration key(s): autoInstal"]
+        );
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:474]: Fixed in 9e816b92. The unknown-key warning test is now wrapped under kakehashi and still asserts the misspelled autoInstal warning while accepting autoInstall.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 4]
```diff
@@ -68,3 +72,99 @@ impl Kakehashi {
...
+    };
+
+    let mut ignored: Vec<_> = object
+        .keys()
+        .filter(|key| {
+            !matches!(
+                key.as_str(),
+                "searchPaths"
+                    | "languages"
+                    | "captureMappings"
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 234ae891. is_known_configuration_key now reads the RawWorkspaceSettings JSON schema properties through crate::config::json_schema(), so recognized didChangeConfiguration keys follow the serde/schemars source of truth instead of a separate hard-coded match list. Added known_configuration_keys_come_from_schema_properties coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 5]
```diff
@@ -68,3 +78,243 @@ impl Kakehashi {
...
+fn parse_normalized_client_configuration(
+    normalized: NormalizedClientConfiguration,
+) -> Result<ClientConfigurationUpdate, serde_json::Error> {
+    let settings = serde_json::from_value::<RawWorkspaceSettings>(normalized.raw_value.clone())?;
+
+    Ok(ClientConfigurationUpdate {
+        settings,
+        warnings: normalized.warnings,
+    })
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:112]: Fixed in 9e816b92. parse_normalized_client_configuration now moves normalized.raw_value into serde_json::from_value instead of cloning it.

[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:112]: Fixed in 9e816b92. This now moves normalized.raw_value into serde_json::from_value, so the extra JSON clone is gone.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 6]
```diff
@@ -68,3 +78,243 @@ impl Kakehashi {
...
+    let ignored = ignored_keys(value);
+    if ignored.is_empty() {
+        Vec::new()
+    } else {
+        vec![format!(
+            "Ignored unknown client configuration key(s): {}",
+            ignored.join(", ")
+        )]
+    }
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:200]: Fixed in 9e816b92. The settings root is normalized first, unrelated settings.* payloads no longer warn, known flat settings-root keys still parse, and wrapped kakehashi objects still merge sibling known keys. Added ignores_other_servers_settings_without_warning and parses_settings_root_known_keys coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 7]
```diff
@@ -68,3 +78,243 @@ impl Kakehashi {
...
+                    "cmd": "rust-analyzer"
+                }
+            }
+        }));
+
+        assert!(
+            crate::config::deprecation::json_uses_deprecated_root_markers(&normalized.raw_value)
+        );
+        assert!(parse_normalized_client_configuration(normalized).is_err());
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:665]: Fixed in 9e816b92. Added ignores_other_servers_settings_without_warning, covering settings.gopls with no kakehashi section and asserting it parses as an empty update without warnings.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 8]
```diff
@@ -68,3 +78,243 @@ impl Kakehashi {
...
+        .get("settings")
+        .and_then(|settings| settings.get("kakehashi"))
+        .or_else(|| value.get("kakehashi"))
+    else {
+        let warnings = ignored_key_warnings(&value);
+        return NormalizedClientConfiguration {
+            raw_value: value,
+            warnings,
+        };
+    };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Closing this one intentionally. A targeted warning whenever settings.kakehashi is absent would still be noisy for valid whole-editor payloads such as settings.gopls. The new behavior is: if settings contains known kakehashi keys, parse them at that root; if it contains only other server/editor sections, treat it as an empty update with no warning. That is covered by parses_settings_root_known_keys and ignores_other_servers_settings_without_warning.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 9]
```diff
@@ -68,3 +78,303 @@ impl Kakehashi {
...
+        let warnings = if has_known_configuration_key(&settings_root) {
+            ignored_key_warnings(&settings_root)
+        } else {
+            Vec::new()
+        };
+        return NormalizedClientConfiguration {
+            raw_value: settings_root,
+            warnings,
+        };
+    };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 6874dc3e. If settings contains only other server sections but the top-level payload has flat Kakehashi keys, normalize_kakehashi_settings now falls back to the top-level flat configuration with wrapper keys removed. Added parses_top_level_known_keys_when_settings_contains_other_servers.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 10]
```diff
@@ -68,3 +81,329 @@ impl Kakehashi {
...
+
+            if is_known_configuration_key(key) {
+                raw_object
+                    .entry(key.clone())
+                    .or_insert_with(|| candidate.clone());
+            } else {
+                warnings.push(key.clone());
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 6874dc3e. Wrapped kakehashi payloads now merge flat siblings from both the top-level object and the settings wrapper object, so settings.kakehashi plus settings.diagnosticsDebounceMs is preserved. Added merges_settings_root_flat_siblings_with_wrapped_settings.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 11]
```diff
@@ -68,3 +81,329 @@ impl Kakehashi {
...
+        .or_else(|| settings_root.get("kakehashi"))
+        .filter(|value| value.is_object())
+        .or_else(|| value.get("kakehashi"))
+        .filter(|value| value.is_object())
+    else {
+        let warnings = if has_known_configuration_key(&settings_root) {
+            ignored_key_warnings(&settings_root)
+        } else {
+            Vec::new()
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 6874dc3e. Unknown scalar/non-object keys that look like flat Kakehashi settings now warn even without a known sibling, so {"autoInstal": false} reports the typo again. Section maps and dotted editor settings stay quiet; added recognizes_empty_client_configuration and ignores_dotted_editor_settings_without_warning coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 12]
```diff
@@ -68,3 +81,439 @@ impl Kakehashi {
...
+            .any(|(key, value)| should_warn_unknown_key(key, value))
+    })
+}
+
+fn should_warn_unknown_key(key: &str, value: &serde_json::Value) -> bool {
+    !is_known_configuration_key(key)
+        && key != "settings"
+        && key != "kakehashi"
+        && !key.contains('.')
+        && !value.is_object()
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in 6ab1e1d2. should_warn_unknown_key now still suppresses lower-case object sections such as gopls, but warns on unknown camelCase object keys that look like Kakehashi configuration typos, e.g. languageServerss. Added warns_about_object_valued_camel_case_typos coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 13]
```diff
@@ -34,13 +38,22 @@ impl Kakehashi {
...
+
         // Merge onto current effective settings (not from scratch).
         // The current settings already reflect defaults < user < project < initializationOptions,
         // so merging preserves languages and other fields set during initialize.
         let current_ts = self.settings_manager.load_raw_settings();
         // SAFETY: merge_workspace_settings(Some, Some) always returns Some, so unwrap_or_return is
         // defensive only — the None branch is unreachable under the current implementation.
-        let Some(merged_ts) = merge_workspace_settings(Some((*current_ts).clone()), Some(parsed))
+        let Some(merged_ts) =
+            merge_workspace_settings(Some((*current_ts).clone()), Some(parsed.settings))
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in d4a89dae. The merge call now uses the destructured settings value directly instead of parsed.settings.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 14]
```diff
@@ -34,13 +38,22 @@ impl Kakehashi {
             }
         };
 
+        for warning in &parsed.warnings {
+            self.notifier().log_warning(warning).await;
+        }
+
+        if raw_workspace_settings_is_empty(&parsed.settings) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in d4a89dae. didChangeConfiguration now destructures the owned ClientConfigurationUpdate, consumes warnings for logging, and keeps settings as the owned value for the later empty check and merge.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 15]
```diff
@@ -68,3 +82,454 @@ impl Kakehashi {
...
+        .filter(|value| value.is_object())
+    else {
+        let top_level_flat = flat_configuration_without_wrappers(&value);
+        let raw_value = if has_configuration_signal(&settings_root) {
+            settings_root
+        } else if has_configuration_signal(&top_level_flat) {
+            top_level_flat
+        } else {
+            settings_root
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in d8334f44. normalize_kakehashi_settings now keeps settings_root borrowed for lookup and only clones it when the no-wrapper path needs to return it as raw_value.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 16]
```diff
@@ -68,3 +82,454 @@ impl Kakehashi {
...
+        let warnings = if has_configuration_signal(&raw_value) {
+            ignored_key_warnings(&raw_value)
+        } else {
+            Vec::new()
+        };
+        return NormalizedClientConfiguration {
+            raw_value,
+            warnings,
+        };
+    };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:154]: Fixed in d8334f44. The no-wrapper path now computes the settings_root and top_level_flat configuration signals once, selects raw_value with the corresponding has_signal flag, and reuses that flag for warning generation.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 17]
```diff
@@ -68,3 +82,453 @@ impl Kakehashi {
...
+    })
+}
+
+fn should_warn_unknown_key(key: &str, value: &serde_json::Value) -> bool {
+    !is_known_configuration_key(key)
+        && key != "settings"
+        && key != "kakehashi"
+        && !key.contains('.')
+        && (!value.is_object() || key.chars().any(|ch| ch.is_ascii_uppercase()))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:278]: Fixed in 841ffca5. Unknown object-valued keys now warn if they are one edit away from a schema-derived configuration key, so `languagess: {}` is reported while unrelated lower-case server sections such as `gopls` remain quiet. Added warns_about_lowercase_object_typos_near_known_keys coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 18]
```diff
@@ -68,3 +82,515 @@ impl Kakehashi {
...
+        if let Some(root_object) = value.as_object() {
+            merge_flat_sibling_settings(raw_object, &mut warnings, root_object);
+        }
+
+        if value.get("settings").is_some()
+            && let Some(settings_object) = settings_root.as_object()
+        {
+            merge_flat_sibling_settings(raw_object, &mut warnings, settings_object);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:167]: Fixed in c08d537b. Wrapped `settings`-root flat siblings are merged before top-level siblings, so they take precedence when both locations provide the same known key, while the inner `kakehashi` object still has highest precedence through `or_insert`. Added settings_root_flat_siblings_precede_top_level_siblings coverage.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 19]
```diff
@@ -68,3 +82,530 @@ impl Kakehashi {
...
+}
+
+fn raw_workspace_settings_is_empty(settings: &RawWorkspaceSettings) -> bool {
+    settings.search_paths.is_none()
+        && settings.languages.is_empty()
+        && settings.capture_mappings.is_empty()
+        && settings.auto_install.is_none()
+        && settings.diagnostics_debounce_ms.is_none()
+        && settings.language_servers.is_none()
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:349]: Fixed in 8aa3b75e. RawWorkspaceSettings now derives equality and owns an is_effectively_empty helper that compares against RawWorkspaceSettings::default(), so future fields participate automatically. didChangeConfiguration now delegates its empty-update check to that method.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 20]
```diff
@@ -68,3 +82,525 @@ impl Kakehashi {
...
+        if value.get("settings").is_some()
+            && let Some(settings_object) = settings_root.as_object()
+        {
+            merge_flat_sibling_settings(raw_object, &mut warnings, settings_object);
+        }
+
+        if let Some(root_object) = value.as_object() {
+            merge_flat_sibling_settings(raw_object, &mut warnings, root_object);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:167]: Fixed in e1aa60b62 by intentionally removing support for the double-wrapper shape. The supported wrappers are top-level `kakehashi` and `settings.kakehashi`; `settings.settings.kakehashi` is outside the client payload contract and is now pinned by a regression test as an empty/no-op update.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 21]
```diff
@@ -68,3 +82,540 @@ impl Kakehashi {
...
+        };
+    };
+
+    let mut warnings = Vec::new();
+    let mut raw_value = inner.clone();
+
+    if let Some(raw_object) = raw_value.as_object_mut() {
+        if value.get("settings").is_some()
+            && let Some(settings_object) = settings_root.as_object()
+        {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in d7c6d01d. The settings sibling lookup now uses `value["settings"].as_object()` as suggested, preserving the same semantics while simplifying the branch.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 22]
```diff
@@ -68,3 +82,540 @@ impl Kakehashi {
...
+        } else if top_level_flat_has_signal {
+            (top_level_flat, true)
+        } else {
+            (settings_root.clone(), false)
+        };
+        let warnings = if has_signal {
+            ignored_key_warnings(&raw_value)
+        } else {
+            Vec::new()
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:149]: Fixed in a532c311. Settings-root payloads still take precedence, but typo-like top-level flat keys dropped by that choice are now included in the ignored-key warning and covered by a regression test.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 23]
```diff
@@ -68,3 +82,538 @@ impl Kakehashi {
...
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs:149]: Fixed in a532c311 as well. This duplicate thread is outdated; settings-root payloads now include ignored top-level flat keys in the warning path, with a regression test for `{ "settings": { "autoInstall": false }, "autoInstal": true }`.

File: src/lsp/lsp_impl/workspace/did_change_configuration.rs [hunk 24]
```diff
@@ -68,3 +82,561 @@ impl Kakehashi {
...
+        let top_level_flat_has_signal = has_configuration_signal(&top_level_flat);
+        let top_level_flat_ignored = ignored_keys(&top_level_flat);
+        let (raw_value, has_signal) = if settings_root_has_signal {
+            (settings_root.clone(), true)
+        } else if top_level_flat_has_signal {
+            (top_level_flat, true)
+        } else {
+            (settings_root.clone(), false)
+        };
+        let warnings = if has_signal {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/workspace/did_change_configuration.rs]: Fixed in bff29d74. When settings-root has the configuration signal, known top-level flat keys are now merged into the selected settings object with settings-root taking precedence; the mixed payload is covered by a regression test.
